src/btree/row_srch.c

```diff
@@ -102,7 +102,18 @@ __wt_search_insert(
     match = skiphigh = skiplow = 0;
     ins = last_ins = NULL;
     for (i = WT_SKIP_MAXDEPTH - 1, insp = &ins_head->head[i]; i >= 0;) {
-        if ((ins = *insp) == NULL) {
+        /*
+         * The algorithm requires that the skip list insert pointer is only read once within the
+         * loop. Whilst the compiler can change the code in a way that it reads the insert pointer
+         * value from memory again in the following code.
+         *
+         * In addition, a CPU with weak memory ordering, such as ARM, may speculatively read an old
+         * value. It is not OK and the reason is explained in the following comment.
+         *
+         * Place a read barrier here to avoid these issues.
+         */
+        WT_ORDERED_READ(ins, *insp);
```

Comment:
Does `*insp` need to be volatile? Otherwise what prevents torn reads (and other fun things)?

We should probably put something like [this](https://elixir.bootlin.com/linux/latest/source/include/asm-generic/rwonce.h#L44) from linux's `READ_ONCE` into `WT_ORDERED_READ` to add `volatile` qualification when doing the write to tell the compiler that it is important that this is done precisely as we say here.

(I am fully aware that `volatile` is not spec-correct for multi-threading, but since we aren't moving to the C11 model in this commit, we should at least do what we can to tell the compiler that this read is "special")

Reply:
> Does *insp need to be volatile? Otherwise what prevents torn reads (and other fun things)?

WiredTiger [requires](https://github.com/wiredtiger/wiredtiger/blob/32ee82a40f6d58579ed28ce5eb979a9fbda391df/src/include/gcc.h#L53) atomic reads and writes of 64-bit memory locations. So we don't take additional steps to ensure atomicity (i.e., prevent torn reads/writes) when loading and storing pointer values to shared memory.

Is there reason to believe that the platforms we run on don't provide this guarantee?

Reply:
Are you sure that you're getting a read every time, and that it's never optimized away?

Reply:
> Are you sure that you're getting a read every time, and that it's never optimized away?

I was responding specifically to the concern about torn reads. So fair point -- using a `volatile` would also prevent "other fun things" by ensuring that we really load a value from memory.

In this case, however, I can't see it. The barrier in the loop ensures that each iteration of the loop must load a new value of `ins` from memory. This code is not updating that information in memory, so the compiler has no way to know what is in that memory without reading it. If the compiler has an oracle that can accurately determine what is in memory at some point between two barriers, then it is giving us a correct value! :-)

I think @RedBeard0531's question about including a `volatile` qualifier in the `WT_ORDERED_READ()` macro is a reasonable question/proposal for what we might want to do in WT in the future. So in this thread I'm basically arguing that it is sufficient and desirable to just make the smallest change that will fix the issue in WT-10461 so we minimize the chances of inadvertently breaking other stuff. Happy to follow up about broader improvements in subsequent tickets.

Reply:
For what it's worth, I went and dug up a 2019 Linux kernel thread specifically talking about torn reads on aligned memory locations.  Where they had seen it, which partially drives the use of volatile in the kernel.

https://lore.kernel.org/all/CAHk-=wj2t+GK+DGQ7Xy6U7zMf72e7Jkxn4_-kGyfH3WFEoH+YQ@mail.gmail.com/T/#u

> We do require a _minimum_ of compiler sanity, but the compiler turning a non-marked read into two reads has definitely happened, and can be very subtle. Even if on a C level it looks like a single access, and correctness doesn't care about _what_ the value we read is, it might be turned by the compiler into two separate accesses that get two different values, and then the end result may be insane and unreliable.

And a link to all the UB we're inviting in with bare loads and stores: https://lwn.net/Articles/793253/

I don't know if it's in this commit, but at some point we've got to move away from relying so much on undefined behavior.

Reply:
Thanks @hanumantmk. That's fascinating! (Also a bit frustrating as I'd really like to know when/where/how they found read tearing here. But Linus is silent on that...)